### ISSUE TITLE
Failure on lektor v3.3.2

### ISSUE THREAD

**Issue by uyar** [sphericalpm/lektor-tinymce#4]
This plugin fails with the following error message on lektor 3.3.2:

```
  File "/home/uyar/.cache/lektor/packages/68475a57fb73d96ddc1c25c948a524eb/lektor_tinymce.py", line 70, in on_server_spawn
    for path, endpoint in dash.endpoints:
AttributeError: module 'lektor.admin.modules.dash' has no attribute 'endpoints'
```

**Comment by uyar** [comment]
@dairiki Could you point me toward a fix for this?

**Comment by dairiki** [comment]
This plugin is using internal, undocumented parts of both Flask and Lektor to inject its own custom javascript into the admin page.  As such, it's not surprising that it broke.

It could be fixed, but the result would be as fragile as it is now.   And (without conditionally checking Lektor version) it would no longer work with earlier version of Lektor (if that matters.)  And it will almost certainly break again once Lektor 3.4.0 is released due to lektor/lektor#987.

I'm not enough of an expert on Flask to know whether there might be a better way to "decorate" a view in a Blueprint.  If there were, that would be the ideal solution.  I've googled a bit and haven't found anything along those lines, however.

----

Note that [lektor-simplemde](https://github.com/sphericalpm/lektor-simplemde) uses the same code to inject its javascript into the admin page and is going to suffer the same issues.

----

I'm not adverse to adding a simple hook to Lektor to enable something like this.  I'm not quite sure what the best API for it may be.

This would likely require creating an `on_webapp_created` Lektor hook that could be used to configure and/or modify the admin app.
The simplest (but most limited) idea would be to add some way to specify extra javascript to be injected on the admin page.
It might be nice to also provide a means whereby plugins can apply custom middleware to the Flask app.  Maybe that's all that's needed — it's not quite as clean but middleware can, of course, inject whatever it wants into the admin page by munging the returned HTML.

**Comment by uyar** [comment]
> This plugin is using internal, undocumented parts of both Flask and Lektor to inject its own custom javascript into the admin page. As such, it's not surprising that it broke.
> 

I figured as much. I've looked into the source code for the error but didn't see an obvious fix.

> It could be fixed, but the result would be as fragile as it is now. And (without conditionally checking Lektor version) it would no longer work with earlier version of Lektor (if that matters.) And it will almost certainly break again once Lektor 3.4.0 is released due to [lektor/lektor#987](https://github.com/lektor/lektor/pull/987).
> 

There are a couple of sites that I run for friends and for them I need an HTML editor. So, I'll have to stay on Lektor 3.3.1 for those.

I'll keep following Lektor releases closely to see if any chance comes up for a nicer solution. Thanks!

**Comment by apatrushev** [comment]
Yeah. The plugin is quite hackish but it is the only way we found to inject custom editors. If you check the `setup.py` you can realise that we are specifying strict lektor versions as dependency. You can also do that with `requirements.txt` for example. We will release support for v3.3.6 soon because it was fixed in #5.

**Comment by apatrushev** [comment]
Please check version 0.4 (already on PyPI) for the latest lektor v3.3.6 support.
